sbtc/src/idpack/codec/mod.rs

```diff
@@ -0,0 +1,190 @@
+use super::{segment, ALLOC_BYTES_LIMIT};
+
+mod decoder;
+mod encoder;
+pub mod strategies;
+
+/// Maximum number of u64 values that can be safely allocated in a single segment.
+/// This limit balances maximum compression efficiency with memory safety,
+/// allowing for ~131K values per segment without risking allocation attacks.
+pub const VALUE_COUNT_LIMIT: u32 = ALLOC_BYTES_LIMIT / 8; // ~131,072 values
+
+/// Size in bytes for encoding flags in the serialized format.
+/// The flag byte packs encoding type and strategy-specific optimizations
+/// for maximum compression efficiency.
+pub const FLAGS_SIZE: usize = 1;
+
+/// Mask for extracting the encoding type from the flags byte (bits 0-1).
+/// The encoding type determines the fundamental compression strategy used.
+#[rustfmt::skip]
+const ENCODING_TYPE_MASK: u8      = 0b0000_0011; // Bits 0-1
+
+/// Single value encoding type (00).
+/// Optimized for isolated values with zero payload bytes.
+#[rustfmt::skip]
+const TYPE_SINGLE: u8             = 0b0000_0000;
+
+/// Bitmap encoding type (01).
+/// Optimal for dense values in small ranges (>25% density).
+#[rustfmt::skip]
+const TYPE_BITSET: u8             = 0b0000_0001;
+
+/// Mask for extracting encoding-specific flags (bits 2-6).
+/// These bits store strategy-specific optimizations to minimize encoded size.
+#[rustfmt::skip]
+const ENCODING_FLAGS_MASK: u8     = 0b0111_1100;
+
+/// Flag 1 (bit position 2) - First encoding-specific optimization flag.
+/// For Bitset: Embedded bitmap flag
+/// For FWDelta: Embedded bit width flag
+#[rustfmt::skip]
+#[allow(unused)]
+const ENCODING_FLAG_1: u8         = 0b0000_0100;
+
+/// Flag 2 (bit position 3) - Second encoding-specific optimization flag.
+/// For Bitset: Embedded length flag
+/// For FWDelta: Tiny sequence flag
+#[rustfmt::skip]
+#[allow(unused)]
+const ENCODING_FLAG_2: u8         = 0b0000_1000;
+
+/// Flag 3 (bit position 4) - Third encoding-specific optimization flag.
+/// For Bitset: First bit of embedded length or bitmap
+/// For FWDelta: First bit of embedded bit width
+#[rustfmt::skip]
+const ENCODING_FLAG_3: u8         = 0b0001_0000;
+
+/// Flag 4 (bit position 5) - Fourth encoding-specific optimization flag.
+/// For Bitset: Second bit of embedded length or bitmap
+/// For FWDelta: Second bit of embedded bit width
+#[rustfmt::skip]
+const ENCODING_FLAG_4: u8         = 0b0010_0000;
+
+/// Flag 5 (bit position 6) - Fifth encoding-specific optimization flag.
+/// For Bitset: Third bit of embedded length or bitmap
+/// For FWDelta: Third bit of embedded bit width
+#[rustfmt::skip]
+const ENCODING_FLAG_5: u8         = 0b0100_0000;
+
+/// Continuation flag (bit position 7), indicates whether additional
+/// segments follow in the serialized format. This enables streaming
+/// compression across multiple segments for maximum efficiency.
+#[rustfmt::skip]
+const FLAG_CONTINUATION: u8       = 0b1000_0000;
```

Comment:
Hmmm, at a glance this seems to add unnecessary complexity. It seems unnecessary since we always know whether we need to keep going or not. Decoding can be simple, continue to decode the next segment if there are more bytes to decode, no need for a flag. Alternatively, after decoding each segment, we know how many numbers we have and how many are expected and we can use that to decide whether to continue decoding. Although the first approach seems best to me, either of them seem like a simplification.